distribution/lib/Standard/Base/0.0.0-dev/src/Random.enso

```diff
@@ -22,6 +25,66 @@ type Random_Number_Generator
     ## A random number generator.
     Value java_random
 
+    ## Gets the next random Boolean.
+    boolean : Decimal
+    boolean self =
+        self.java_random.nextBoolean
+
+    ## Gets the next random Decimal between 0 and 1.
+    decimal : Decimal
+    decimal self =
+        self.java_random.nextDouble
+
+    ## Gets the next random Decimal from a normal distribution with mean 0 and std-dev 1.
+    gaussian : Decimal
+    gaussian self =
+        self.java_random.nextGaussian
+
+    ## Gets the next random integer between min (inclusive) and max (exclusive).
+
+       Arguments:
+        - min: the minimum value (inclusive) of the random integer.
+        - max: the maximum value (exclusive) of the random integer.
+    integer : Integer -> Integer -> Integer
+    integer self min=0 max=100 =
+        range = max - min
+        if range < Java_Integer.MAX_VALUE then min + (self.java_random.nextInt range) else
+            min + (self.decimal * range).floor
+
+## Gets the next random Boolean.
+
+   Arguments:
+    - generator: the random number generator to use.
+boolean : Random_Number_Generator -> Decimal
+boolean generator=(Random_Number_Generator.new) =
+    generator.boolean
```

Comment:
Is it definitely good practice to create a _new_ generator on _each_ call by default?

Do we surely want to have these statics instead of requiring the user to create a new generator node and using that?

Reply:
I'm especially worried that with how this is done currently, on _every_ recomputation in the GUI, the result will be different. This may lead to some weird behaviour in the GUI, no?

Although I realise the problem is even worse, even if I did the following in the GUI:
```
gen = Random_Number_Generator.new 42
x1 = gen.boolean
x2 = gen.integer
x3 = gen.integer
op1 = if x1 then x2+x3 else x2-x3
op2 = x2*10
```

Depending on what is recomputed and in what order, I can still get nondeterministic results, because for example `x2` may for some reason get recomputed but `gen` and `x1` not, so it will not get a resetted generator and will return a different value.

I'm not an expert in how the GUI is recomputing stuff (may be worth to cc @4e6 here?), but looking how partial recomputations may happen, I think the approach to randomness we have here may be problematic for the GUI.

I guess we may keep that for now, but I think we should create a task to redesign how we are approaching randomness in Enso. If we want it to play well with the GUI, I don't think we can rely on mutable objects, but instead use something closer to how our `State` works so that the order of generating each random object is tied with each other. And we may want to use something similar to `in Input` context for instantiating the RNG with a random seed so that it is not recomputed every time and for example for the whole GUI session the initial seed stays FROZEN and is only recomputed upon explicit request or on restart. That's just a bunch of suggestions, but the main point is - I think we need to revise how we approach random generators before we start using them more in the GUI.

Reply:
I agree there is a risk non-deterministic ness here but I think this is still a good API and approach. The current noise does not produce good random values. In general, I think users are less concerned on the deterministic value than they are the ability to make a random number.

I will replace the new with Java cached singleton.

Reply:
I think the only way to make it Enso-GUI friendly is to thread some generator value through, probably by State.

Still, why do we need a singleton and instead cannot rely on the user creating an RNG? I'd prefer that as I think it will be closer to what the final GUI-friendly API may look like, so it may be better to get users accustomed to it from the get-go instead of changing it significantly later on.

Reply:
The non-deterministcness is especially problematic if the user wants determinism and uses a constant seed - due to how GUI is recomputing values, unless I'm mistaken, it is possible only for subgraphs to be recomputed and thus still getting nondeterministic results. This is quite bad as if the user is using a constant seed they are going to have some determinism expectations. I'm not saying it's an issue that needs solving right now, but I think we should keep in mind to revisit it somewhere in the (preferable nearer than further) future.

Reply:
Ok will remove the static methods but this feels a lot less friendly to a beginner end user.

Reply:
I'm open to further discussion if needed, but I think the static-oriented approach will be hard to get right once we have a proper randomness solution. Although maybe it is possible if we are able to define some Random monad, I'm just not sure what our options will be in the future.

Reply:
Removed them - I do think a data tool should have them but not going to argue.

Reply:
Sorry if I'm pushing my viewpoint too much - I think that you are right about the statics, my point was that threading through the generator _may_ make it safer for the GUI - but overall it may actually require an overall redesign of how we handle randomness in the GUI to make it work well, so as a temporary solution I'm not really sure if one is much better than the other.